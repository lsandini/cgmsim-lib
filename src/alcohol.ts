import { TreatmentDelta, Treatment, TreatmentDrug, GenderType } from './Types';
import { getDrugActivity } from './drug';
import logger, { getDeltaMinutes, getTreatmentActivity } from './utils';

function getAlcoholActivity(
	gender: 'Male' | 'Female',
	weightKg: number,
	peakMin: number,
	durationMin: number,
	timeMin: number, //minutesAgo
	units: number,
): number {
	const peakAlcoholAmountMin = 60; //min
	const r = gender === 'Male' ? 0.68 : 0.55;
	const eliminationRate = gender === 'Male' ? 0.016 / 60 : 0.018 / 60; //g/100ml/min
	const peakAmount = ((units * 12) / (weightKg * 1000 * r)) * 100; //g/100ml
	const washoutDuration = peakAlcoholAmountMin + peakAmount / eliminationRate; //min
	const unitsWeighted = units * 12 * (80 / weightKg);
	if (washoutDuration < timeMin) {
		return (
			getTreatmentActivity(
				peakMin,
				durationMin,
				timeMin - washoutDuration,
				unitsWeighted,
			) / 35
		);
	}
	return 0;
}

export const computeAlcoholActivity = (
	treatments: TreatmentDelta[],
	weight: number,
	gender: GenderType,
) => {
	const treatmentsActivity = treatments.map((e) => {
		const minutesAgo = e.minutesAgo;
		const units = e.units;
		const activity = getAlcoholActivity(
			gender,
			weight,
			e.peak,
			e.duration,
			minutesAgo,
			units,
		);
		return activity;
	});
	logger.debug('these are the last Alcohol: %o', treatmentsActivity);
	const resultAct = treatmentsActivity.reduce((tot, activity) => {
		return tot + activity;
	}, 0);
	return resultAct;
};

export default function (
	treatments: TreatmentDrug[],
	weight: number,
	gender: GenderType,
): number {
	//Find Alcohol boluses

	const lastALC = getDrugActivity(treatments, weight, 'ALC');
	const activityALC =
		lastALC.length > 0 ? computeAlcoholActivity(lastALC, weight, gender) : 0;
	logger.debug('these are the last ALC: %o', { lastALC, activityALC });

	return activityALC / 2;
}
