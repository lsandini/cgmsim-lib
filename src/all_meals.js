const moment = require('moment');
module.exports = function(treatments = []) {
  const now = moment();
  const meals = treatments
    .filter(e => e.carbs && moment(e.mills).diff(now, 'minutes') > -360)
    .map(e => ({
      ...e,
      time:  (Date.now() - moment(e.mills).valueOf()) / (1000 * 60)
    }));

  console.log('Last 6 hours meal:', meals);
  return meals;
};
