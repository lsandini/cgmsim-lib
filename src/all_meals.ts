import logger from './utils';
import { Treatment, TreatmentDelta } from './Types';
import { getDeltaMinutes } from './utils';

//const logger = pino();

export default function (treatments: Treatment[]): TreatmentDelta[] {

	const meals = treatments
		.filter(e => e.carbs && getDeltaMinutes(e.mills) > 360)
		.map(e => ({
			...e,
			minutesAgo: getDeltaMinutes(e.mills),
		}));

	logger.info('Last 6 hours meal:', meals);
	return meals;
}

