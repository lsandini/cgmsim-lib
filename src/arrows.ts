import logger from './utils';
import { Sgv, Direction } from './Types';

//const logger = pino();


export default function (sgvs: Pick<Sgv, 'sgv'>[]): { sgvdir: number; direction: Direction }[] {
	const arrows = [];
	if (sgvs && sgvs.length >= 4) {
		const sgvdir1 = sgvs[0].sgv - sgvs[1].sgv;
		const sgvdir2 = sgvs[1].sgv - sgvs[2].sgv;
		const sgvdir3 = sgvs[2].sgv - sgvs[3].sgv;
		const sgvdir15min = (sgvdir1 + sgvdir2 + sgvdir3) / 3;
		logger.info('this is the mean SGV 5 min variation in the last 15 minutes: %o', sgvdir15min, 'mg/dl');

		if (sgvdir15min < -10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'DoubleDown'
			});

		} else if (sgvdir15min < -6) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'SingleDown'
			});

		} else if (sgvdir15min < -2) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'FortyFiveDown'
			});

		} else if (sgvdir15min < 2) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'Flat'
			});

		} else if (sgvdir15min < 6) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'FortyFiveUp'
			});

		} else if (sgvdir15min < 10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'SingleUp'
			});

		} else if (sgvdir15min >= 10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'DoubleUp'
			});
		}

	} else {
		arrows.push({
			sgvdir: 0,
			direction: 'Flat'
		});
	}
	logger.info(arrows);

	return arrows;
}