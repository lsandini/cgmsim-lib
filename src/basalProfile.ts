import moment = require('moment');
import { Profile } from './Types';
export default function (profile: Profile[]): number {
	let lastProfile: Profile = null;
	let lastProfiles = profile.sort((first, second) =>
		moment(second.startDate).diff(-moment(first.startDate)),
	);

	if (!lastProfiles || lastProfiles.length === 0) {
		return 0;
	}
	lastProfile = lastProfiles[0];
	const defaultProfile = lastProfile.defaultProfile;
	const defaultProfileBasals = Array.isArray(
		lastProfile.store[defaultProfile].basal,
	)
		? lastProfile.store[defaultProfile].basal[0].value
		: lastProfile.store[defaultProfile].basal;
	return defaultProfileBasals / 60;
}
