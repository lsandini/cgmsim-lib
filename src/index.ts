import simulator from './CGMSIMsimulator';
import simulatorUVA from './UVAsimulator';
import downloads from './downloads';
import uploadEntries from './upload-entries';
import uploadActivity from './upload-activity';
import loadActivity from './load-activity';
import uploadNotes from './upload-notes';
import arrows from './arrows';


export { arrows, simulator, downloads, uploadEntries, uploadActivity, loadActivity, simulatorUVA, uploadNotes };
