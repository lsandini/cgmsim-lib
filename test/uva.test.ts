import moment = require('moment');
import { Profile, Treatment } from 'src/Types';
import simulatorUVA from '../src/UVAsimulator';
beforeAll(() => {
	jest.useFakeTimers('modern');
	jest.setSystemTime(new Date('2022-05-01T11:00:00'));
});

afterAll(() => {
	jest.useRealTimers();
});

describe('uva test default PATIENT', () => {
	test('basal  0.75 from TREATMENTS should generate flat sgv', () => {
		//current treatments generate 0.75U/h
		const treatments = [{
			"_id": "627548f4a3dc0ad67616ac95",
			"eventType": "Announcement",
			"notes": "gla 30",
			"utcOffset": 0,
			"isAnnouncement": true,
			"protein": "",
			"fat": "",
			"duration": 0,
			"profile": "",
			"enteredBy": "Boss",
			"created_at": "2022-05-01T08:41:00",
			"carbs": null,
			"insulin": null
		},];
		let lastState = null;
		const yList = [];
		for (let i = 0; i < 12; i++) {

			const { x, y } = simulatorUVA({ env: { WEIGHT: '80' }, lastState, treatments, profile: [] });
			yList.push(y.Gp);
			lastState = x;
		}

		expect(lastState).toMatchSnapshot()
		expect(yList).toMatchSnapshot()

	})

	test('basal 0.75 from PROFILE should generate flat sgv', () => {

		let lastState = null;
		const yList = [];
		const profile: Profile[] = [{
			defaultProfile: 'pippo',
			startDate: '2022-01-01',
			store: {
				'pippo': {
					basal: 0.75
				}
			}
		}, {
			defaultProfile: 'pippo2',
			startDate: '2021-01-01',
			store: {
				'pippo2': {
					basal: 0.85
				}
			}
		}]
		for (let i = 0; i < 12; i++) {
			const { x, y } = simulatorUVA({ env: { WEIGHT: '80' }, lastState, treatments: [], profile });
			yList.push(y.Gp);
			lastState = x;
		}

		expect(lastState).toMatchSnapshot()
		expect(yList).toMatchSnapshot()
	})

	test('basal 0.75 from PROFILE + 50g CARBS should generate a curve', () => {

		let lastState = null;
		const yList = [];
		const profile: Profile[] = [{
			defaultProfile: 'pippo',
			startDate: '2022-01-01',
			store: {
				'pippo': {
					basal: 0.75
				}
			}
		}, {
			defaultProfile: 'pippo2',
			startDate: '2021-01-01',
			store: {
				'pippo2': {
					basal: 0.85
				}
			}
		}]
		const treatments: Treatment[] = [{
			carbs: 50,
			created_at: '2022-05-01T10:59:00',
		}]

		const now = moment('2022-05-01T11:00:00')
		let yMax = 0;
		for (let i = 0; i < 36; i++) {
			const { x, y } = simulatorUVA({ env: { WEIGHT: '80' }, lastState, treatments, profile });
			yList.push(y.Gp);
			lastState = x;
			yMax = y.Gp > yMax ? y.Gp : yMax;
			jest.setSystemTime(new Date(now.add(5, 'minutes').toISOString()));
		}


		const yAfter1h = yList[11];
		const yAfter2h = yList[23];
		const yAfter3h = yList[35];

		expect(yAfter1h).toBeLessThan(189)
		expect(yAfter1h).toBeGreaterThan(188)

		expect(yAfter2h).toBeLessThan(235)
		expect(yAfter2h).toBeGreaterThan(234)

		expect(yAfter3h).toBeLessThan(228)
		expect(yAfter3h).toBeGreaterThan(227)
		
		expect(yMax).toBeLessThan(234.9)
		expect(yMax).toBeGreaterThan(234.8)
		
		
		expect(lastState).toMatchSnapshot()
		expect(yList).toMatchSnapshot()
	})

	test('basal 0.75 from PROFILE + 50g CARBS + 5U should generate a curve', () => {

		let lastState = null;
		const yList = [];
		const profile: Profile[] = [{
			defaultProfile: 'pippo',
			startDate: '2022-01-01',
			store: {
				'pippo': {
					basal: 0.75
				}
			}
		}]
		const treatments: Treatment[] = [{
			carbs: 50,
			created_at: '2022-05-01T11:30:00',
		},{
			carbs: 0,
			insulin: 7.5,
			created_at: '2022-05-01T11:15:00',
		}]

		const now = moment('2022-05-01T11:00:00')
		let yMax = 0;
		for (let i = 0; i < 36; i++) {
			const { x, y } = simulatorUVA({ env: { WEIGHT: '80' }, lastState, treatments, profile });
			yList.push(y.Gp);
			lastState = x;
			yMax = y.Gp > yMax ? y.Gp : yMax;
			jest.setSystemTime(new Date(now.add(5, 'minutes').toISOString()));
		}


		const yAfter1h = yList[11];
		const yAfter2h = yList[23];
		const yAfter3h = yList[35];

		expect(yAfter1h).toBeLessThan(119)
		expect(yAfter1h).toBeGreaterThan(118)

		expect(yAfter2h).toBeLessThan(164)
		expect(yAfter2h).toBeGreaterThan(163)

		expect(yAfter3h).toBeLessThan(124)
		expect(yAfter3h).toBeGreaterThan(123)
		
		expect(yMax).toBeLessThan(165.6)
		expect(yMax).toBeGreaterThan(165.5)
		
		
		expect(lastState).toMatchSnapshot()
		expect(yList).toMatchSnapshot()
	})

})