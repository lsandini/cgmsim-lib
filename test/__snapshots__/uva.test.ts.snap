// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uva test default PATIENT basal  with gla 30 should generate flat sgv 1`] = `
Object {
  "Gp": 172.16999363425157,
  "Gt": 129.36710057913723,
  "I_": 41.14442017759129,
  "Il": 2.003990855337264,
  "Ip": 2.0572694126441236,
  "Isc1": 55.718153844257,
  "Isc2": 50.207567162989385,
  "Qgut": 0,
  "Qsto1": 0,
  "Qsto2": 0,
  "W": 0,
  "X": 3.6306029496153656,
  "XL": 41.136798965515716,
  "Y": 0,
  "Z": 0,
}
`;

exports[`uva test default PATIENT basal  with gla 30 should generate flat sgv 2`] = `
Array [
  100.00004322991686,
  100.00063492353028,
  100.00250189626965,
  100.00593128856623,
  100.01078205750811,
  100.01666662064541,
  100.02309014171705,
  100.0295325805898,
  100.03549230499027,
  100.04050767395145,
  100.04416666216194,
  100.04611008372085,
  100.04603137865107,
  100.04367454087162,
  100.038831046713,
  100.03133626887137,
  100.02106566209093,
  100.0079308973572,
  99.99187605773699,
  99.97287396968255,
  99.95092271781185,
  99.92604237335155,
  99.89827195365011,
  99.8676666208005,
  99.83429512048515,
  99.79823745705112,
  99.7595827971124,
  99.71842759135265,
  99.6748739024279,
  99.62902792576206,
  99.5809986894409,
  99.53089691922526,
  99.4788340548246,
  99.42492140392375,
  99.36926942097155,
  99.31198709837709,
  99.25318145847004,
  99.19295713534116,
  99.13141603645782,
  99.06865707473256,
  99.00477596249046,
  98.93986505952898,
  98.874013268178,
  98.80730596894692,
  98.73982499098422,
  98.67164861217178,
  98.60285158423002,
  98.53350517872221,
  98.4636772503155,
  98.39343231408644,
  98.32283163405013,
  98.25193332044589,
  98.18079243363499,
  98.10946109275176,
  98.03798858751037,
  97.96642149179964,
  97.89480377790431,
  97.82317693037456,
  97.75158005872632,
  97.68005000829865,
  97.60862146871827,
  97.5373270795314,
  97.46619753265912,
  97.39526167141297,
  97.3245465858815,
  97.25407770455755,
  97.18387888212902,
  97.11397248340091,
  97.04437946335194,
  96.97511944336169,
  96.90621078366898,
  96.83767065214292,
  96.7695150894652,
  96.70175907083494,
  96.63441656431722,
  96.56750058596371,
  96.50102325183946,
  96.43499582709119,
  96.36942877219582,
  96.30433178652639,
  96.23971384937164,
  96.17558325854382,
  96.11194766670592,
  96.04881411554557,
  95.9861890679199,
  95.92407843808986,
  95.86248762015903,
  95.8014215148259,
  95.74088455455517,
  95.68088072726606,
  95.62141359863392,
  95.56248633309325,
  95.50410171362756,
  95.4462621604253,
  95.38896974847749,
  95.33222622418715,
  95.2760330210573,
  95.22039127451895,
  95.16530183595795,
  95.11076528599406,
  95.05678194706363,
  95.00335189535245,
  94.95047497212309,
  94.89815079447743,
  94.8463787655921,
  94.79515808446234,
  94.74448775518688,
  94.69436659582371,
  94.6447932468453,
  94.59576617921861,
  94.54728370213424,
  94.49934397040658,
  94.4519449915656,
  94.40508463265867,
  94.35876062678074,
  94.31297057934735,
  94.26771197412715,
  94.22298217904604,
  94.17877845177605,
  94.13509794512082,
  94.09193771220758,
  94.04929471149647,
  94.00716581161504,
  93.96554779602776,
  93.92443736754669,
  93.88383115269161,
  93.84372570590578,
  93.80411751363316,
  93.76500299826324,
  93.72637852194828,
  93.68824039029786,
  93.65058485595571,
  93.61340812206211,
  93.57670634560678,
  93.54047564067514,
  93.5047120815916,
  93.46941170596352,
  93.4345705176278,
  93.40018448950413,
  93.36624956635681,
  93.33276166746806,
  93.29971668922462,
  93.26711050762101,
  93.23493898068054,
  93.2031979507966,
  93.17188324699664,
  93.14099068712999,
  93.11051607998176,
  93.08045522731493,
  93.05080392584154,
  93.02155796912554,
  92.9927131494183,
  92.96426525942861,
  92.93621009402885,
  92.90854345189842,
  92.88126113710616,
  92.85435896063355,
  92.8278327418391,
  92.80167830986676,
  92.77589150499796,
  92.75046817995036,
  92.72540420112318,
  92.70069544979151,
  92.67633782324991,
  92.65232723590735,
  92.62865962033419,
  92.60533092826248,
  92.58233713154122,
  92.55967422304681,
  92.53733821755117,
  92.51532515254712,
  92.49363108903366,
  92.47225211226083,
  92.45118433243651,
  92.43042388539524,
  92.40996693323069,
  92.38980966489257,
  92.36994829674896,
  92.35037907311523,
  92.33109826675053,
  92.31210217932231,
  92.2933871418409,
  92.27494951506365,
  92.2567856898709,
  92.23889208761368,
  92.2212651604346,
  92.20390139156262,
  92.18679729558225,
  92.16994941867868,
  92.15335433885878,
  92.13700866614982,
  92.1209090427755,
  92.10505214331094,
  92.08943467481723,
  92.07405337695563,
  92.05890502208327,
  92.04398641532991,
  92.02929439465697,
  92.01482583089975,
  92.00057762779278,
  91.98654672197982,
  91.9727300830081,
  91.95912471330833,
  91.94572764816051,
  91.93253595564647,
  91.91954673658921,
  91.90675712448008,
  91.89416428539411,
  91.88176541789403,
  91.8695577529234,
  91.85753855368932,
  91.84570511553548,
  91.83405476580587,
  91.82258486369903,
  91.81129280011439,
  91.80017599748992,
  91.78923190963285,
  91.77845802154205,
  91.76785184922447,
  91.75741093950428,
  91.74713286982636,
  91.73701524805367,
  91.727055712259,
  91.71725193051195,
  91.70760160066054,
  91.69810245010872,
  91.6887522355894,
  91.6795487429333,
  91.67048978683466,
  91.66157321061281,
  91.65279688597126,
  91.64415871275327,
  91.63565661869518,
  91.62728855917702,
  91.61905251697097,
  91.61094650198781,
  91.6029685510215,
  91.59511672749233,
  91.58738912118845,
  91.57978384800616,
]
`;

exports[`uva test default PATIENT basal 0.70 from PROFILE should generate flat sgv 1`] = `
Object {
  "Gp": 186.79258235613025,
  "Gt": 141.43742162834323,
  "I_": 37.776007480446765,
  "Il": 1.827759122422773,
  "Ip": 1.8888017206357495,
  "Isc1": 51.28205128204566,
  "Isc2": 46.21020005621574,
  "Qgut": 0,
  "Qsto1": 0,
  "Qsto2": 0,
  "W": 0,
  "X": 0.2612491440926876,
  "XL": 37.77575534490338,
  "Y": 0,
  "Z": 0,
}
`;

exports[`uva test default PATIENT basal 0.70 from PROFILE should generate flat sgv 2`] = `
Array [
  100.00004443966257,
  100.00066279085601,
  100.00266472677394,
  100.00647965292272,
  100.01215279062865,
  100.01950984994552,
  100.0282814763839,
  100.03817343912634,
  100.04890127344669,
  100.06020541051817,
  100.07185642672918,
  100.08365555820106,
  100.09543308312013,
  100.10704584710473,
  100.11837454088408,
  100.12932101251374,
  100.13980573782939,
  100.14976549681077,
  100.15915126754811,
  100.1679263328666,
  100.1760645873536,
  100.18354902969773,
  100.19037042455558,
  100.1965261184764,
  100.20201899518909,
  100.2068565565255,
  100.2110501162832,
  100.21461409535772,
  100.2175654074674,
  100.21992292573617,
  100.22170702128383,
  100.2229391657998,
  100.22364159084127,
  100.22383699730356,
  100.22354830916171,
  100.22279846617825,
  100.22161025081903,
  100.2200061451166,
  100.218008213674,
  100.21563800941419,
  100.21291649905288,
  100.20986400561065,
  100.20650016558483,
  100.2028438986757,
  100.19891338820884,
  100.19472607061749,
  100.19029863254632,
  100.18564701431679,
  100.18078641865218,
  100.17573132370215,
  100.1704954995328,
  100.16509202735878,
  100.15953332089495,
  100.15383114929145,
  100.14799666119357,
  100.1420404095367,
  100.13597237674638,
  100.1298020000656,
  100.12353819677894,
  100.11718938914154,
  100.11076352885797,
  100.10426812098449,
  100.09771024715637,
  100.09109658806315,
  100.08443344511481,
  100.0777267612589,
  100.07098214092184,
  100.06420486906043,
  100.05739992931885,
  100.0505720212961,
  100.04372557693387,
  100.0368647760423,
  100.02999356098405,
  100.02311565054248,
  100.01623455299983,
  100.00935357845655,
  100.00247585042133,
  99.99560431670507,
  99.98874175964959,
  99.98189080572492,
  99.97505393452681,
  99.96823348720646,
  99.96143167436372,
  99.95465058343443,
  99.94789218560112,
  99.94115834225603,
  99.93445081104359,
  99.9277712515087,
  99.92112123037623,
  99.91450222648544,
  99.90791563540208,
  99.90136277373001,
  99.89484488314248,
  99.88836313415241,
  99.88191862963998,
  99.87551240815424,
  99.86914544700512,
  99.86281866516079,
  99.85653292596402,
  99.85028903968126,
  99.8440877658963,
  99.83792981575982,
  99.83181585410594,
  99.82574650144508,
  99.8197223358427,
  99.8137438946922,
  99.80781167638985,
  99.80192614191913,
  99.79608771635138,
  99.7902967902684,
  99.7845537211135,
  99.7788588344756,
  99.77321242531175,
  99.76761475911238,
  99.76206607301327,
  99.7565665768584,
  99.75111645421637,
  99.74571586335496,
  99.74036493817522,
  99.73506378910875,
  99.72981250398043,
  99.72461114883853,
  99.71945976875483,
  99.71435838859595,
  99.70930701376832,
  99.70430563093754,
  99.69935420872459,
  99.69445269837917,
  99.68960103443193,
  99.68479913532677,
  99.6800469040339,
  99.6753442286446,
  99.6706909829488,
  99.66608702699583,
  99.6615322076395,
  99.65702635906766,
  99.6525693033174,
  99.6481608507757,
  99.64380080066674,
  99.63948894152591,
  99.635225051661,
  99.631008899601,
  99.62684024453317,
  99.62271883672786,
  99.61864441795257,
  99.61461672187448,
  99.61063547445232,
  99.60670039431783,
  99.60281119314672,
  99.59896757601963,
  99.59516924177323,
  99.59141588334181,
  99.58770718808917,
  99.58404283813147,
  99.58042251065089,
  99.5768458782006,
  99.57331260900081,
  99.56982236722624,
  99.56637481328556,
  99.56296960409232,
  99.55960639332788,
  99.55628483169639,
  99.55300456717228,
  99.54976524523984,
  99.54656650912555,
  99.54340800002305,
  99.54028935731091,
  99.53721021876353,
  99.53417022075496,
  99.53116899845617,
  99.52820618602554,
  99.52528141679332,
  99.52239432343907,
  99.51954453816359,
  99.51673169285432,
  99.5139554192453,
  99.51121534907072,
  99.50851111421336,
  99.50584234684733,
  99.50320867957521,
  99.50060974556027,
  99.49804517865323,
  99.49551461351416,
  99.49301768572944,
  99.49055403192392,
  99.48812328986834,
  99.4857250985825,
  99.48335909843365,
  99.48102493123068,
  99.47872224031418,
  99.47645067064248,
  99.47420986887332,
  99.47199948344216,
  99.46981916463652,
  99.46766856466634,
  99.46554733773137,
  99.46345514008463,
  99.46139163009278,
  99.45935646829311,
  99.45734931744745,
  99.45536984259282,
  99.45341771108941,
  99.45149259266552,
  99.44959415945962,
  99.44772208605974,
  99.44587604954052,
  99.44405572949725,
  99.44226080807783,
  99.44049097001219,
  99.43874590263948,
  99.43702529593295,
  99.43532884252286,
  99.4336562377171,
  99.43200717952003,
  99.43038136864914,
  99.4287785085501,
  99.42719830540993,
  99.42564046816848,
  99.4241047085279,
  99.4225907409612,
  99.42109828271855,
  99.41962705383257,
  99.41817677712207,
  99.41674717819429,
  99.41533798544589,
  99.41394893006287,
  99.412579746019,
  99.41123017007324,
  99.40989994176597,
  99.40858880341429,
  99.40729650010601,
  99.40602277969315,
  99.40476739278398,
  99.40353009273444,
  99.40231063563881,
  99.40110878031925,
  99.3999242883148,
  99.39875692386974,
  99.39760645392083,
  99.39647264808445,
  99.39535527864258,
  99.39425412052856,
  99.39316895131206,
  99.39209955118365,
  99.39104570293875,
  99.39000719196123,
  99.38898380620637,
  99.38797533618362,
  99.38698157493883,
  99.3860023180361,
  99.38503736353935,
  99.38408651199344,
  99.38314956640522,
  99.38222633222396,
  99.38131661732197,
  99.38042023197447,
  99.37953698883956,
  99.37866670293792,
  99.3778091916325,
  99.3769642746075,
  99.3761317738479,
  99.37531151361833,
  99.37450332044202,
  99.37370702307965,
  99.37292245250826,
  99.37214944189974,
  99.37138782659933,
  99.37063744410446,
  99.36989813404301,
  99.36916973815205,
  99.36845210025609,
  99.36774506624555,
  99.36704848405543,
  99.36636220364363,
  99.36568607696961,
  99.36501995797265,
  99.36436370255078,
  99.36371716853915,
  99.36308021568888,
  99.36245270564582,
  99.36183450192925,
  99.36122546991093,
  99.36062547679391,
  99.36003439159185,
  99.35945208510802,
  99.35887842991472,
  99.35831330033248,
  99.35775657240971,
]
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 10U should generate a curve with min 40 1`] = `
Object {
  "Gp": 80.04810422214628,
  "Gt": 40.28697248590791,
  "I_": 208.42877476055517,
  "Il": 44.76743838707893,
  "Ip": 19.99316654766212,
  "Isc1": 176.40956260401776,
  "Isc2": 254.25793769144948,
  "Qgut": 0,
  "Qsto1": 0,
  "Qsto2": 0,
  "W": 0,
  "X": 338.01703372002186,
  "XL": 87.38616640758718,
  "Y": 0,
  "Z": 0,
}
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 10U should generate a curve with min 40 2`] = `
Array [
  100.00004264161005,
  100.0006213714681,
  100.00242271052365,
  99.99792384675492,
  99.90673296498714,
  99.59909800024302,
  98.96215339822214,
  97.91534156519907,
  96.41177823388135,
  94.43520509098101,
  91.9954279300371,
  89.12335312196704,
  85.86608792575532,
  82.2823120235828,
  78.43802110801597,
  74.40270126465836,
  70.245978842596,
  66.03478650703819,
  61.83108190049705,
  57.69014483856612,
  53.65945990503258,
  49.77816473540069,
  46.07701388000602,
  42.57877884156717,
  40,
  40,
  40,
  40,
  40,
  40,
  40,
  40,
  40,
  40,
  40,
  40,
]
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 50g CARBS + 5U should generate a curve 1`] = `
Object {
  "Gp": 237.71160477794848,
  "Gt": 164.04949378844253,
  "I_": 139.04962760701002,
  "Il": 6.451626007441896,
  "Ip": 5.589723471118134,
  "Isc1": 84.810779553499,
  "Isc2": 129.8627622012441,
  "Qgut": 1176.0809941175373,
  "Qsto1": 13.363407253453056,
  "Qsto2": 1026.1103826643443,
  "W": 0,
  "X": 109.21648810360648,
  "XL": 96.18453805076754,
  "Y": 0,
  "Z": 0,
}
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 50g CARBS + 5U should generate a curve 2`] = `
Array [
  100.00004344074452,
  100.00063978009236,
  100.00253027350361,
  100.00022145358407,
  99.93352251449454,
  99.70549349667368,
  99.36967894376468,
  100.12142808950391,
  103.10274182257979,
  108.28238455043054,
  115.10910078366179,
  122.90477989974627,
  131.03493344770885,
  138.97454040663996,
  146.32505688416595,
  152.80894987326482,
  158.25447083286465,
  162.57702855942043,
  165.760420781662,
  167.83955847117625,
  168.88541457348984,
  168.99242331502975,
  168.2682845298544,
  166.8259925843457,
  164.77784871243082,
  162.2311904881151,
  159.28556339940982,
  156.03106083916649,
  152.5475704853508,
  148.90468787540897,
  145.16209051040914,
  141.37020408761003,
  137.57103153082514,
  133.79905117742422,
  130.0821202677148,
  126.44234296699388,
]
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 50g CARBS should generate a curve 1`] = `
Object {
  "Gp": 432.6247096665882,
  "Gt": 340.46068896209886,
  "I_": 38.57724584323862,
  "Il": 1.940889689379203,
  "Ip": 1.998094193876495,
  "Isc1": 54.75534609101646,
  "Isc2": 48.7799616573907,
  "Qgut": 480.2344473447339,
  "Qsto1": 2.505580166758789,
  "Qsto2": 400.20308230431124,
  "W": 0,
  "X": 2.0828726555780763,
  "XL": 37.82546520360286,
  "Y": 0,
  "Z": 0,
}
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 50g CARBS should generate a curve 2`] = `
Array [
  100.13725958657677,
  101.66671679221494,
  105.76653288723423,
  112.42478387876761,
  121.10152166645324,
  131.12795040323837,
  141.8782273587602,
  152.83298434722747,
  163.59258009768376,
  173.8683851820691,
  183.46456829131233,
  192.2480304575594,
  200.15852177726111,
  207.1822441277268,
  213.33515276775458,
  218.65386189514217,
  223.18842688918403,
  226.99680439514734,
  230.1407204863062,
  232.6826763233896,
  234.6838488240752,
  236.2026811005261,
  237.29399475996456,
  238.00848980906625,
  238.39252662459808,
  238.48810819955128,
  238.3330000986862,
  237.96094088512334,
  237.40190785256956,
  236.68241230710095,
  235.82580590329954,
  234.85258508686033,
  233.78068488924404,
  232.6257564527501,
  231.40142497292766,
  230.119526418398,
]
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 500g CARBS should generate a curve with max 400 1`] = `
Object {
  "Gp": 707.55656999924,
  "Gt": 452.78248957686606,
  "I_": 37.63501554830254,
  "Il": 1.8712006820591789,
  "Ip": 1.9328330272679328,
  "Isc1": 53.96902765394085,
  "Isc2": 47.19081523736483,
  "Qgut": 29204.67595082871,
  "Qsto1": 6082.262534675929,
  "Qsto2": 41798.94434707729,
  "W": 0,
  "X": 0.5224347663442177,
  "XL": 37.49672892692031,
  "Y": 0,
  "Z": 0,
}
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE + 500g CARBS should generate a curve with max 400 2`] = `
Array [
  100.00004344074452,
  100.00063978009236,
  100.00253027350361,
  100.00602685486191,
  100.01102094323277,
  100.0171621292742,
  100.43564333859274,
  105.02927190399573,
  117.32987768907162,
  137.3006307191245,
  163.32314847852822,
  193.38954209090733,
  225.58199479759315,
  258.3540533640414,
  290.5296655468007,
  321.26333383006187,
  349.9882539271432,
  376.3598776591702,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
  400,
]
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE with 60% exercise should generate curve sgv 1`] = `
Object {
  "Gp": 118.84153617067572,
  "Gt": 81.43288793763907,
  "I_": 38.57724584323862,
  "Il": 1.940889689379203,
  "Ip": 1.998094193876495,
  "Isc1": 54.75534609101646,
  "Isc2": 48.7799616573907,
  "Qgut": 0,
  "Qsto1": 0,
  "Qsto2": 0,
  "W": 5342.999999999982,
  "X": 2.0828726555780763,
  "XL": 37.82546520360286,
  "Y": 8.438303689132207e-9,
  "Z": 0.9040418973768898,
}
`;

exports[`uva test default PATIENT basal 0.75 from PROFILE with 60% exercise should generate curve sgv 2`] = `
Array [
  99.89814037698908,
  99.4225317645088,
  98.58079249584156,
  97.47135416720675,
  96.17665561268514,
  94.75021735382575,
  93.22412631325132,
  91.61755128436114,
  89.94259150223087,
  88.20770959415093,
  86.4195990416085,
  84.58415529460736,
  82.70696437894097,
  80.8789522430743,
  79.32098317247905,
  78.02148826733692,
  76.89899992135366,
  75.88736703334253,
  74.94517142388875,
  74.04903931300358,
  73.1864320113355,
  72.35076967925912,
  71.53859680949273,
  70.74812546887541,
  69.97946226704421,
  69.23512726465486,
  68.51711471735496,
  67.82605116816852,
  67.16181712859564,
  66.52395017459978,
  65.91183709895937,
  65.32480318833764,
  64.76215342253506,
  64.22319131797187,
  63.70722742426185,
  63.21358306950837,
]
`;
